### ISSUE TITLE
[PATCH] Support for detecting libraries in several directories simultaneously (migrated from Trac #993)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2546]
Original ticket http://projects.scipy.org/numpy/ticket/993
Reported 2009-02-02 by atmention:FrancescAlted, assigned to atmention:rgommers.

I've found a problem with `numpy.distutils` when I want to specify multiple libraries for a !NumPy component that are spread in two or more directories.  For example, for linking against MKL, a `site.cfg` with the next contents:

```
[mkl]
library_dirs = /opt/intel/Compiler/11.0/074/mkl/lib/em64t/:/opt/intel/Compiler/11.0/074/lib/intel64
include_dirs =  /opt/intel/Compiler/11.0/074/mkl/include/
mkl_libs = mkl_gf_lp64, mkl_gnu_thread, mkl_core, iomp5
```

gives the next output:

```
mkl_info:
  libraries mkl_gf_lp64,mkl_gnu_thread,mkl_core,iomp5 not found in /opt/intel/Compiler/11.0/074/mkl/lib/em64t/
  libraries mkl_gf_lp64,mkl_gnu_thread,mkl_core,iomp5 not found in /opt/intel/Compiler/11.0/074/lib/intel64
  NOT AVAILABLE
```

I've tracked down the problem in that `numpy.distutils` tries to find the complete set of libraries in one single directory, instead in the complete list.

I'm attaching a fix that is able to find the set of required libraries in all the directories specified in `library_dirs` and only in one of them.  With it, the output is:

```
mkl_info:
  FOUND:
    libraries = ['mkl_gf_lp64', 'mkl_gnu_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/opt/intel/Compiler/11.0/074/mkl/lib/em64t/', '/opt/intel/Compiler/11.0/074/lib/intel64']
    define_macros = [('SCIPY_MKL_H', None)]
    include_dirs = ['/opt/intel/Compiler/11.0/074/mkl/include/']
```

So, it seems to work well. Moreover, the fix simplifies quite a bit the logic in `system_info.py`.  However, caveat emptor: as I've only tested the MKL section, more testing should be carried out with other libraries than MKL before to apply the patch.

Finally, attention should be payed to the next issue:

In `atlas_info.calc_info()` method, I've not completely simplified the loop:

```
        for d in lib_dirs:
            atlas = self.check_libs2(d,atlas_libs,[])
            lapack_atlas = self.check_libs2(d,['lapack_atlas'],[])
            if atlas is not None:
                lib_dirs2 = [d] + self.combine_paths(d,['atlas*','ATLAS*'])
                lapack = self.check_libs2(lib_dirs2,lapack_libs,[])
                if lapack is not None:
                    break
            if atlas:
                atlas_1 = atlas
```

because I'm not sure about how can affect the use of `self.combine_paths()` in the previous loop, but I'd say that it could be rewritten as:

```
        atlas = self.check_libs2(lib_dirs,atlas_libs,[])
        lapack_atlas = self.check_libs2(lib_dirs,['lapack_atlas'],[])
        if atlas is not None:
            lib_dirs2 = [lib_dirs] + self.combine_paths(lib_dirs,['atlas*','ATLAS*'])
            lapack = self.check_libs2(lib_dirs2,lapack_libs,[])
            atlas_1 = atlas
```

So, if anybody with more knowledge than me can double check the patch, it would be great.

Thanks

**Comment by thouis** [comment]
Attachment in Trac by atmention:FrancescAlted, 2009-02-02: [system_info.patch](http://projects.scipy.org/numpy/attachment/ticket/993/system_info.patch)

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-02-20

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-03-02

I am bit worried about this patch - it adds more complexity to system_info, which is already a monstrosity. Given that the problem it solves can be dealt with some links, I prefer not merging it.

**Comment by thouis** [comment]
Comment in Trac by atmention:FrancescAlted, 2009-03-02

Well, while I agree that applying it few days before a release is not a good idea, I still think it offers a real value for users (i.e. avoid to deal with links in library files).

IMHO, it would be better if this ticket gets deferred instead of closed definitely.  Applying it in trunk after releasing 1.3 would be good enough to get it tested by developers.  In case it poses grave problems, I'd happily close it then.

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-03-02

To explain a bit better: I am worried about merging it at all, not just for 1.3.0. The whole concept used in system_info does not make much sense, we should have proper library check instead. Adding more code to this only makes future transition harder IMHO.

**Comment by thouis** [comment]
Comment in Trac by atmention:FrancescAlted, 2009-03-02

I see your point, but the patch in fact simplifies some of the logic in system_info, and does not add any significant complexity, so I won't be afraid of applying it.

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-03-02

Ok, I will look at it for 1.4.0 then.

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2009-03-10

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2010-04-24

David, what do you think now?

**Comment by thouis** [comment]
Comment in Trac by trac user nicopinto, 2011-02-13

Any chance of getting this merged soon? Thanks!

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2011-02-14

David, it seems like modifying the site.cfg is a better alternative (from the user's perspective) than creating symlinks (especially on platforms such as Windows?).  The patch could possibly still be simplified further to give the desired behavior.  While I buy your argument of a maintenance burden, counter-intuitive setup behavior causes users to complain to the list, which isn't much better.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-03-24

Ralf, could you take a look at this and make a judgement call about adding it? To be in 1.6, it would ideally have been applied long ago, but there seems to be interest from more than one person in having it applied.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-24

It's certainly too late to go into 1.6, but it may be a good idea to put it in master for a prolonged workout. I'll try to find some time to test this, then come back to it. 

Still would like to get David's OK for it, he knows that code much better than I do.

**Comment by thouis** [comment]
Comment in Trac by trac user npinto, 2012-03-06

Bump ;-)

Any way to see this coming to 1.7 ? Let me know how can I help.

Regards,

N

**Comment by thouis** [comment]
Comment in Trac by trac user npinto, 2012-03-06

The patch still works after a quick fix, available here:
https://gist.github.com/1983883

**Comment by thouis** [comment]
Comment in Trac by trac user npinto, 2012-03-06

I followed up with a PR on github:
https://github.com/numpy/numpy/pull/229

let me know what you thinks

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-03-14

PR merged in commit:a4dbfc1

Thanks all!

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-03-16

This introduced a bug, which was fixed in commit:83480b6
